docs/ExtensibleMaps.md

```diff
@@ -0,0 +1,133 @@
+NMR, similar to existing program and hook providers.
+
+**New NMR Provider Interface**: `EBPF_MAP_INFO_EXTENSION_IID`
+
+**Provider Registration Data**:
+```c
+typedef struct _ebpf_map_provider_data {
+    ebpf_extension_header_t header;
+    uint32_t custom_map_type;  // Single map type per provider
+    uint32_t base_map_type;
+    ebpf_map_provider_dispatch_table_t* dispatch_table;
+} ebpf_map_provider_data_t;
+```
+
+**Provider Dispatch Table**:
+```c
+typedef struct _ebpf_map_provider_dispatch_table {
+    ebpf_extension_header_t header;
+    ebpf_map_create_notify_t create_map_notify;
+    ebpf_map_delete_notify_t delete_map_notify;
+    ebpf_map_associate_program_type_t associate_program;
+    ebpf_map_find_element_notify_t find_element_notify;
+    ebpf_map_update_element_notify_t update_element_notify;
+    ebpf_map_delete_element_notify_t delete_element_notify;
+    ebpf_map_get_next_key_and_value_notify_t get_next_key_and_value_notify;
+} ebpf_map_provider_dispatch_table_t;
+```
+
+An extension (provider) needs to provide the above dispatch table. eBPF runtime will invoke the above functions in the
+following scenarios:
+1. **Map Creation** -- eBPF runtime will invoke create_map_notify to validate the key and value sizes, and optionally get actual value size.
+2. **Map Deletion** -- eBPF runtime will invoke delete_map_notify to notify the extension that the map is being deleted.
+3. **CRUD operation** -- For each CRUD operation, eBPF runtime will invoke the corresponding dispatch function to notify extension. In case of update functions, extension can optionally provide a different value to be stored in the map.
+
+**Client Services** (provided by eBPF core):
+```c
+typedef struct _ebpf_map_client_dispatch_table {
+    ebpf_extension_header_t header;
+    ebpf_map_find_element_t find_element_function;
+} ebpf_map_client_dispatch_table_t;
+```
+eBPF runtime will expose a *find_element_function* dispatch function that extension can use to query a map value, given the key.
+
+## Map Type Enum Partitioning
+Currently map type IDs are allocated from a global namespace. With extensible maps, global map type ID space is
+partitioned into 2 disjoint sets: for global maps (implemented in eBPFCore) and for the extensible maps.
+Global maps will use IDs from 1 to 4095. Extensible maps will use IDs 4096 onwards.
+
+- **Global Maps (1-4095)**: Reserved for core eBPF runtime map types (hash, array, etc.)
+- **Extensible Maps (4096+)**: Available for extension-implemented custom map types
+
+Extensions *can* reserve unique map type IDs by submitting PRs to update the enum in the eBPF repository.
+
+## Map Discovery and Creation
+
+**Dynamic Provider Discovery**: Uses NMR's built-in discovery mechanism
+- No central registry required - providers are discovered on-demand during map creation
+- When a map is created with an extensible type (>= 4096), eBPF core:
+  1. Creates an `ebpf_extensible_map_t` structure with NMR client characteristics
```

Comment:
I think the term "extensible map" is incorrect (based on the definition in this document) and should not be used anywhere.